assemble.go

```diff
@@ -119,6 +200,22 @@ func AssembleFile(ctx context.Context, name string, idx Index, s Store, seeds []
 					continue
 				}
 				c := job.segment.chunks()[0]
+
+				// If we already took this chunk from the store we can reuse it by looking
+				// into the selfseed.
+				if segment := ss.getChunk(c.ID); segment != nil {
+					copied, cloned, err := segment.WriteInto(f, c.Start, c.Size, blocksize, isBlank)
+					if err != nil {
+						return err
+					}
+					stats.addBytesCopied(copied)
+					stats.addBytesCloned(cloned)
+					// Record this segment's been written in the self-seed to make it
+					// available going forward
+					ss.add(job.segment)
+					continue
+				}
```

Comment:
Hmm, so this won't work quite right since `c` is only the first chunk in a segment to be copied. It'll likely produce invalid file since if there's a segment of 2 chunks to be copied, and the first of those is in the file already, then this will only copy the first chunk and the 2nd one will be skipped, creating a broken output file.

Basically, the introduction of the "plan" has made the self-seed very much useless since it's impossible to use it in the plan. It only gets populated at runtime and it's not currently possible to predict when the selfseed is written, so it can't be part of the plan.

Perhaps the selfseed needs to be retired completely. I wouldn't expect that to have much of a perf impact, especially when a local cache is used.

Reply:
You could actually save the selfseed if `getChunk()` takes a whole segment (not just the first chunk) and then returns an equal segment from itself if one exists of the same length with the same chunks.

Reply:
I changed this to `getSegment()` and basically implemented what you suggested.

Reply:
> Hmm, so this won't work quite right since c is only the first chunk in a segment to be copied. It'll likely produce invalid file since if there's a segment of 2 chunks to be copied, and the first of those is in the file already, then this will only copy the first chunk and the 2nd one will be skipped, creating a broken output file.

I was thinking that this was not failing because when we create a Plan, a segment has more than one chunk only if we are getting it from a seed (and selfseed is not in that list). Like when we get a single chunk from the store.

Reply:
Probably we should do an assert on the number of chunks that we have? Like if we don't have a seed segment we assert that we just have a single chunk.

If that's not the case we would write a single chunk when reading it from the output file (the `isBlank` portion) and also when we get a it from the store. And if the segment had more than one chunk, the others will be lost.

Reply:
So you can actually simplify the selfseed implementation to operate on single chunks, since it's never going to be part of a plan anyway, and won't need the "longestMatch" functionality if I'm reading this right.

Reply:
Right. I reverted to the previous `getChunk()` and added a comment to explain this assumption and also a check with `panic()` if that's not the case. If we trip there, there is a programming error.